security/advancedtls/provider_test.go

```diff
@@ -0,0 +1,261 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
+	"math/big"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc/credentials/tls/certprovider"
+	"google.golang.org/grpc/security/advancedtls/testdata"
+)
+
```

Comment:
@dfawley and I have one question for the test. Because in this PR we are testing some file reloading behaviors, we set the reloading interval to 500 millisecond,  and then use a tmp file for credential contents. We first use the content from maybe a file called credential_1, and wait for 1 second, then change the content to credential_2, and check the changes are picked up by the Provider(s).
Does that testing logic sound good to you? I am mainly concerned about using the tmp file, but couldn't think of a better way for this. Also since the picking up logic in Provider is done in a separate goroutine, theoretically speaking there is only a weak guarantee(500ms < 1s) that the changes will be updated. Would that cause problems?

Reply:
What you're describing is more like an end-to-end test.  Usually this is my preferred way of doing testing, but when they involve loading files and sleeps, they have the potential to be flaky.  Also, sleeps tend to add up and result in slow test runtimes when you start expanding tests.

It would be preferable to inject a function to "read a file" that provides data directly from memory in the tests.  And instead of a sleep, using a channel for synchronization is preferable, so it would also be better to inject a "sleep" function that simply unblocks when the test is ready in the tests.  Is this possible here?

Reply:
I've thought over this and couldn't come up with a very good solution. We probably could do the injection by separating the reading and sleeping parts out and use an interface for us to be able to inject different logic, but that also affects the readability of the code, and even worse, we need to add one more interface parameter in public API telling which implementation we'd like to use.
I think the difficulty stems from the fact that we want two different implementations for test and real code, so we have to use some ways to tell the application which to use at some point. We probably can benefit from [Wire](https://github.com/google/wire) or [go mock](https://github.com/golang/mock), but I think we still have to abstract out the interface and add parameters. May I know if you have any suggestions? And by the way can we use wire or go mock in grpc-go?

Reply:
Define a couple of functions used in your code as vars:

```
var (
  readKeyPair =  tls.LoadX509KeyPair
  newTicker = time.NewTicker
)
```

And in your code, use `readKeyPair` instead of `tls.LoadX509KeyPair` and `newTicker` instead of `time.NewTicker`.

Override `readKeyPair` to return contents of a cert in your test.
Override `newTicker` to return a ticker with a small timer. Thinking about it, you might not even have to override the ticker functionality since you can option to pass the `Interval`.

Reply:
Ah yeah, thanks for this neat suggestion! I haven't thought about reassigning those functions. Going this direction seems to be very promising :)